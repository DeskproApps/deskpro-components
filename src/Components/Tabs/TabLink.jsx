import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'utils/noop';

/**
 * Renders a link in a tab group.
 */
const TabLink = ({ name, active, onClick, children, className, ...props }) => {
  const classes = classNames(
    'dp-tabs__tab',
    className,
    {
      'dp-tabs__tab--active': active
    }
  );

  return (
    <li className={classes} {...props} onClick={() => { onClick(name); }}>
      {children}
    </li>
  );
};

TabLink.propTypes = {
  /**
   * The name of the tab.
   */
  name:      PropTypes.string.isRequired,
  /**
   * Whether or not the tab is active.
   */
  active:    PropTypes.bool,
  /**
   * CSS classes to apply to the element.
   */
  className: PropTypes.string,
  /**
   * Children to render.
   */
  children:  PropTypes.node,
  /**
   * Called when the link is clicked. Receives the name of the tab.
   */
  onClick:   PropTypes.func
};

TabLink.defaultProps = {
  active:  false,
  onClick: noop
};

export default TabLink;
