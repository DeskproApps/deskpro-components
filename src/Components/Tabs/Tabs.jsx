import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'utils/noop';
import { objectKeyFilter } from 'utils/objects';

/**
 * Renders a group of links as tabs.
 */
export default class Tabs extends React.Component {
  static propTypes = {
    /**
     * Name of the active tab.
     */
    active:    PropTypes.string.isRequired,
    /**
     * CSS classes to apply to the element.
     */
    className: PropTypes.string,
    /**
     * Children to render.
     */
    children:  PropTypes.node.isRequired,
    /**
     * Called when the active tab changes. Receives the name of the tab.
     */
    onChange:  PropTypes.func
  };

  static defaultProps = {
    active:    '',
    className: '',
    onChange:  noop,
  };

  handleClick = (active) => {
    if (active !== this.props.active) {
      this.props.onChange(active);
    }
  };

  render() {
    const { className, children, active, ...props } = this.props;

    const tabs = React.Children.map(children, child => React.cloneElement(child, {
      onClick: this.handleClick,
      active:  child.props.name === active
    }));

    return (
      <ul className={classNames('dp-tabs', className)} {...objectKeyFilter(props, Tabs.propTypes)}>
        {tabs}
      </ul>
    );
  }
}
