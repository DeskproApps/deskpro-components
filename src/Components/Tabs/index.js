import Tabs from './Tabs';
import TabLink from './TabLink';

export {
  Tabs,
  TabLink,
};
