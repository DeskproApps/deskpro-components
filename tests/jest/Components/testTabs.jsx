import React from 'react';
import renderer from 'react-test-renderer';
import { Tabs, TabLink } from '../../../src/Components/Tabs/index';

describe('>>> Tabs --- Snapshot', () => {
  it('+++capturing Snapshot of TabLink', () => {
    const renderedValue = renderer.create(<TabLink name="one">
        Tab One
    </TabLink>).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
  it('+++capturing Snapshot of Tabs', () => {
    const renderedValue = renderer.create(<Tabs active="one">
      <TabLink name="one">
          Tab One
      </TabLink>
      <TabLink name="two">
          Tab Two
      </TabLink>
    </Tabs>).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
