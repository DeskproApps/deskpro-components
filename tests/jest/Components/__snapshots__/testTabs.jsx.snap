// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Tabs --- Snapshot +++capturing Snapshot of TabLink 1`] = `
<li
  className="dp-tabs__tab"
>
  <div
    onClick={[Function]}
  >
    Tab One
  </div>
</li>
`;

exports[`>>> Tabs --- Snapshot +++capturing Snapshot of Tabs 1`] = `
<ul
  className="dp-tabs"
>
  <li
    className="dp-tabs__tab dp-tabs__tab--active"
  >
    <div
      onClick={[Function]}
    >
      Tab One
    </div>
  </li>
  <li
    className="dp-tabs__tab"
  >
    <div
      onClick={[Function]}
    >
      Tab Two
    </div>
  </li>
</ul>
`;
